File: apps/macos/Sources/CornerTasks/Crypto/Identity.swift
import CryptoKit
import Foundation

enum IdentityError: Error {
    case badDid
    case verificationFailed
}

struct Identity {
    let privateKey: Curve25519.Signing.PrivateKey
    let publicKey: Curve25519.Signing.PublicKey
    let accountDid: String
    /// methodSpecificId is the part of the DID after `did:key:` (e.g. `z6Mk...`).
    let methodSpecificId: String

    /// Build an Identity from a 32-byte seed (the output of HKDF for identity).
    static func fromIdentitySeed(_ seed: Data) throws -> Identity {
        let priv = try Curve25519.Signing.PrivateKey(rawRepresentation: seed)
        return make(privateKey: priv)
    }

    /// Convenience: build from the BIP-39 seed.
    static func fromBip39Seed(_ bip39Seed: Data) throws -> Identity {
        try fromIdentitySeed(DataKey.identitySeed(from: bip39Seed))
    }

    private static func make(privateKey: Curve25519.Signing.PrivateKey) -> Identity {
        let pub = privateKey.publicKey
        let did = encodeDid(publicKey: pub.rawRepresentation)
        let msi = String(did.dropFirst("did:key:".count))
        return Identity(privateKey: privateKey, publicKey: pub, accountDid: did, methodSpecificId: msi)
    }

    /// `did:key` encoding for an Ed25519 public key (multicodec 0xed 0x01, base58btc, multibase prefix `z`).
    static func encodeDid(publicKey: Data) -> String {
        var prefixed = Data([0xed, 0x01])
        prefixed.append(publicKey)
        return "did:key:z" + Base58.encode(prefixed)
    }

    /// Recover the 32-byte Ed25519 public key from a `did:key` string.
    static func publicKey(fromDid did: String) throws -> Data {
        guard did.hasPrefix("did:key:z") else { throw IdentityError.badDid }
        let body = String(did.dropFirst("did:key:z".count))
        guard let raw = Base58.decode(body) else { throw IdentityError.badDid }
        guard raw.count == 34, raw[0] == 0xed, raw[1] == 0x01 else { throw IdentityError.badDid }
        return raw.suffix(32)
    }

    func sign(_ message: Data) throws -> Data {
        try privateKey.signature(for: message)
    }

    static func verify(signature: Data, message: Data, did: String) throws -> Bool {
        let pkBytes = try publicKey(fromDid: did)
        let pub = try Curve25519.Signing.PublicKey(rawRepresentation: pkBytes)
        return pub.isValidSignature(signature, for: message)
    }

    /// Build a compact JWS DID-JWT per `docs/sync-protocol.md` §8.2. `exp = iat + 300`.
    func makeDidJwt(audience: String, nonce: String, iat: Int) throws -> String {
        let kid = "\(accountDid)#\(methodSpecificId)"
        let header = "{\"alg\":\"EdDSA\",\"typ\":\"JWT\",\"kid\":\"\(kid)\"}"
        let exp = iat + 300
        let payload = "{\"iss\":\"\(accountDid)\",\"sub\":\"\(accountDid)\",\"aud\":\"\(audience)\",\"nonce\":\"\(nonce)\",\"iat\":\(iat),\"exp\":\(exp)}"
        let h = Base64Url.encode(Data(header.utf8))
        let p = Base64Url.encode(Data(payload.utf8))
        let signingInput = "\(h).\(p)"
        let sig = try sign(Data(signingInput.utf8))
        return "\(signingInput).\(Base64Url.encode(sig))"
    }
}

File: apps/web/src/crypto/identity.ts
// Ed25519 identity, sign/verify, and DID-JWT (compact JWS) per docs/sync-protocol.md §8.2.

import * as ed from '@noble/ed25519';
import { didFromPublicKey, methodSpecificId, publicKeyFromDid } from './did';
import { identitySeed } from './keys';

export interface Identity {
  privateKey: Uint8Array;
  publicKey: Uint8Array;
  accountDid: string;
  methodSpecificId: string;
  sign(message: Uint8Array): Promise<Uint8Array>;
  makeDidJwt(args: { audience: string; nonce: string; iat?: number }): Promise<string>;
}

export const fromIdentitySeed = async (seed: Uint8Array): Promise<Identity> => {
  if (seed.length !== 32) throw new Error('identity seed must be 32 bytes');
  const privateKey = new Uint8Array(seed);
  const publicKey = await ed.getPublicKeyAsync(privateKey);
  const accountDid = didFromPublicKey(publicKey);
  const msi = methodSpecificId(accountDid);
  return {
    privateKey,
    publicKey,
    accountDid,
    methodSpecificId: msi,
    sign: (message) => ed.signAsync(message, privateKey),
    makeDidJwt: ({ audience, nonce, iat }) =>
      makeDidJwt({ accountDid, methodSpecificId: msi, privateKey }, { audience, nonce, iat }),
  };
};

export const fromBip39Seed = async (bip39Seed: Uint8Array): Promise<Identity> => {
  const seed = await identitySeed(bip39Seed);
  return fromIdentitySeed(seed);
};

export const verify = async (signature: Uint8Array, message: Uint8Array, did: string): Promise<boolean> => {
  const pub = publicKeyFromDid(did);
  return ed.verifyAsync(signature, message, pub);
};

const b64url = (bytes: Uint8Array): string => {
  let s = '';
  for (const b of bytes) s += String.fromCharCode(b);
  return btoa(s).replaceAll('+', '-').replaceAll('/', '_').replace(/=+$/, '');
};

const b64urlString = (text: string): string => b64url(new TextEncoder().encode(text));

export const makeDidJwt = async (
  id: { accountDid: string; methodSpecificId: string; privateKey: Uint8Array },
  args: { audience: string; nonce: string; iat?: number }
): Promise<string> => {
  const iat = args.iat ?? Math.floor(Date.now() / 1000);
  const exp = iat + 300;
  const kid = `${id.accountDid}#${id.methodSpecificId}`;
  // Field order matches the macOS Identity.makeDidJwt(...) implementation so the
  // base64url-encoded segments are byte-identical.
  const header = `{"alg":"EdDSA","typ":"JWT","kid":"${kid}"}`;
  const payload = `{"iss":"${id.accountDid}","sub":"${id.accountDid}","aud":"${args.audience}","nonce":"${args.nonce}","iat":${iat},"exp":${exp}}`;
  const signingInput = `${b64urlString(header)}.${b64urlString(payload)}`;
  const sig = await ed.signAsync(new TextEncoder().encode(signingInput), id.privateKey);
  return `${signingInput}.${b64url(sig)}`;
};
